src/ratelimiter.rs

```diff
@@ -0,0 +1,140 @@
+use std::time::{Duration, Instant};
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    available_bytes: u64,
+    bytes_per_second: u64,
+    last_update: Instant,
+}
+
+impl RateLimiter {
+    pub fn new(mbps_target: u64, now: Instant) -> Self {
+        let bps_target = mbps_target * 1_000_000;
+        RateLimiter {
+            available_bytes: bps_target,
+            bytes_per_second: bps_target,
+            last_update: now,
+        }
+    }
+
+    pub fn bytes_available(&self, now: Instant) -> u64 {
+        let elapsed = now - self.last_update;
+        let new_bytes = elapsed.as_secs_f32() * self.bytes_per_second as f32;
+        std::cmp::min(
+            self.available_bytes + new_bytes as u64,
+            self.bytes_per_second,
+        )
```

Comment:
This implicitly sets the bucket capacity at 1 second × bytes-per-second. If the chunk size is larger than one second of data, it means there will never be enough `bytes_available()` to send. (At least not in the current approach where we don’t adjust the chunk size.) Maybe it doesn’t happen in this case because the minimum is 1 MB/s and the chunk size is smaller than that, but then somebody later changes the chunk size or switches the limiter unit to kB/s and then it subtly breaks. I would add an explicit `capacity_bytes` field and initialize it to `MAX_CHUNK_LEN`.

Reply:
much better now, thanks!